core/src/executor/validate.rs

```diff
@@ -74,39 +79,93 @@ impl UniqueConstraint {
-        ColumnValidation::SpecifiedColumns(column_defs, specified_columns) => {
-            fetch_specified_unique_columns(&column_defs, &specified_columns)
+    enum Columns {
+        /// key index
+        PrimaryKeyOnly(usize),
+        /// `[(key_index, table_name)]`
+        All(Vec<(usize, String)>),
+    }
+
+    let columns = match &column_validation {
+        ColumnValidation::All(column_defs) => {
+            let primary_key_index = column_defs
+                .iter()
+                .enumerate()
+                .find(|(_, column_def)| {
+                    column_def
+                        .options
+                        .iter()
+                        .any(|ColumnOptionDef { option, .. }| {
+                            matches!(option, ColumnOption::Unique { is_primary: true })
+                        })
+                })
+                .map(|(i, _)| i);
+            let other_unique_column_def_count = column_defs
+                .iter()
+                .filter(|column_def| {
+                    column_def
+                        .options
+                        .iter()
+                        .any(|ColumnOptionDef { option, .. }| {
+                            matches!(option, ColumnOption::Unique { is_primary: false })
+                        })
+                })
+                .count();
+
+            match (primary_key_index, other_unique_column_def_count) {
+                (Some(primary_key_index), 0) => Columns::PrimaryKeyOnly(primary_key_index),
+                _ => Columns::All(fetch_all_unique_columns(column_defs)),
+            }
         }
+        ColumnValidation::SpecifiedColumns(column_defs, specified_columns) => Columns::All(
+            fetch_specified_unique_columns(column_defs, specified_columns),
+        ),
     };
 
-    let unique_constraints: Vec<_> = create_unique_constraints(columns, row_iter)?.into();
-    if unique_constraints.is_empty() {
-        return Ok(());
-    }
-
-    let unique_constraints = Rc::new(unique_constraints);
-    storage.scan_data(table_name).await?.try_for_each(|result| {
-        let (_, row) = result?;
-        Rc::clone(&unique_constraints)
-            .iter()
-            .try_for_each(|constraint| {
-                let col_idx = constraint.column_index;
-                let val = row
-                    .get_value(col_idx)
-                    .ok_or(ValidateError::ConflictOnStorageColumnIndex(col_idx))?;
+    match columns {
+        Columns::PrimaryKeyOnly(primary_key_index) => {
+            for primary_key in row_iter.filter_map(|row| match row.0.get(primary_key_index) {
+                Some(value) => Key::try_from(value).map(Some).transpose(),
+                None => None,
+            }) {
+                let key = primary_key?;
+
+                if storage.fetch_data(table_name, &key).await?.is_some() {
+                    return Err(ValidateError::DuplicateEntryOnPrimaryKeyField(key).into());
+                }
+            }
```

Comment:
[futures::join](https://docs.rs/futures/latest/futures/macro.join.html) can be used to improve concurrency.